### PR TITLE
Feature/41 register page

### PR DESCRIPTION
**Pull request by Heyjp**

### CODE REVIEW

File: client/src/Auth/Register/register.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                    <input type="text" placeholder="Username" className="form-control mb-2" value={props.inputs.username} onChange={(e) => {props.handleChange('username', e)} } />
+                    <label>Password</label>
+                    <input type="password" placeholder="Password" className="form-control mb-2" value={props.inputs.password} onChange={(e) => {props.handleChange('password', e)} }/>
+                    { props.inputs.password && 
+                        <div>
+                            <label>Confirm Password</label>
+                            <input type="password" placeholder="Confirm Password" className="form-control mb-2" value={props.inputs.confirmPassword} onChange={(e) => {props.handleChange('confirmPassword', e)} }/>
+                        </div>
+                    }
+                    <button disabled={(props.inputs.username.length < 2) && (props.inputs.password !== props.inputs.confirmPassword)} className="btn btn-outline-primary" onClick={props.handleSubmit}>Submit</button>
```
Comment (sfiquet) [inline client/src/Auth/Register/register.js]: That should be an `||`, not an `&&`. The button is almost never disabled.

File: client/src/Auth/authContainer.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    handleChange = (input, e) => {
+        this.setState({
+            [input]: e.target.value
+        })
+    }
+
+    handleSubmit = () => {
+
+        axios.post('/api/register')
```
Comment (sfiquet) [inline client/src/Auth/authContainer.js]: How are you passing the data to the api?

File: client/src/Auth/Register/register.js (hunk 1 above)
Comment (Heyjp) [reply client/src/Auth/Register/register.js]: No, it is definitely intended to be an `&&` - for people registering we want both the username to be 3 or more characters and also we want them to have a password which they can make sure matches.  However I could add more prompts in order to inform the user what is necessary to register.

File: client/src/Auth/authContainer.js (hunk 2 above)
Comment (Heyjp) [reply client/src/Auth/authContainer.js]: Since, we didn't have the registration up I left the object blank, but you are right, i'll place it in the api call.

File: client/src/Auth/Register/register.js (hunk 1 above)
Comment (sfiquet) [reply client/src/Auth/Register/register.js]: You are testing on the opposite. Your code says: disable this button if both conditions are true:
- the username is less than 2 (!) chars
AND
- the config password doesn't match the password

What you actually want to do is disable the button when either:
- the username is less than 3 char
AND/OR
- the passwords don't match

Definitely an OR.

And yes, definitely tell the user what is wrong. When the passwords don't match but are the same length, the user can't spot it.

Comment (Heyjp) [reply client/src/Auth/Register/register.js]: My mistakes because we wanted to evaluate them both to `false` however I changed the way I did it to be more manageable hopefully.

File: client/package.json [hunk 3]
```diff
@@ -15,7 +16,7 @@
   "scripts": {
     "build-css": "node-sass-chokidar src/ -o src/",
     "watch-css": "npm run build-css && node-sass-chokidar src/ -o src/ --watch --recursive",
-    "start": "react-scripts start",
+    "start": "npm run build-css && react-scripts start",
```
Comment (sfiquet) [inline client/package.json]: Not sure that's necessary. We should be running watch-css all the time anyway. Unless it's to ensure the css is built before starting the app? My concern is it might be a little too easy to forget to run watch-css and then waste time wondering why the changes to the scss are not applied.

File: client/src/Auth/Register/register.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
+import React from 'react';
+// https://www.flaticon.com/free-icon/circular-avatar_108331
+import avatar from './avatar.png';
+import './register.css';
+
+// Checks whether state values meets valid criteria
+function validateInputs ({ username, password, confirmPassword }) {
+    
+    return {
+        username: username.length < 4,
```
Comment (sfiquet) [inline client/src/Auth/Register/register.js]: Don't use magic numbers in the code. That's how you end up with discrepancies (e.g. password min length is 5 in the code, should be 7 according to error message). Define const variables at the top of the file, e.g. `const MIN-LENGTH-USERNAME = 4;`, and use them in the code like this: `username: username.length < MIN-LENGTH-USERNAME`.

File: client/src/Auth/Register/register.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        password: password.length < 5,
+        confirmPassword: confirmPassword !== password
+    }
+}
+
+// If an input is not focused has value but not complete returns false
+function checkDirty (focus, {username, password, confirmPassword}, errors) {
+
+    return {
+        username: focus !== 'username' && errors.username && username.length > 1,
```
Comment (sfiquet) [inline client/src/Auth/Register/register.js]: Why > 1? Shouldn't that be > 0 or >= 1? Why should the behaviour be different when there is a single character?

File: client/src/Auth/Register/register.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    const isDirty = checkDirty(focus, props.inputs, errors);
+    const isEnabled = Object.keys(errors).some(e => errors[e]);
+
+    return (
+        <div className="container-fluid d-flex flex-column my-auto">
+                <div className="d-flex flex-column col-lg-5 col-md-3 col-sm-8 mx-auto justify-content-center register-wrapper">
+                    <div className="align-center pt-2 mb-3">
+                        <h2 className="register-text">Register</h2>
+                    </div>
+                    <img src={avatar} className="align-self-center my-3" width="120px" />
```
Comment (sfiquet) [inline client/src/Auth/Register/register.js]: You have an ESLing warning here: It needs an alt attribute, probably empty as it's just to pretty up the page.

File: client/package.json (hunk 3 above)
Comment (Heyjp) [reply client/package.json]: Exactly, you will always have an up to date version of your scss every time you run the server. I do not see this as a negative thing.

File: client/src/Auth/Register/register.test.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import {mount, shallow} from 'enzyme';
+
+import Register from './register';
+
+describe('Register', () => {
+  it('username should match inputs.username', () => {
```
Comment (sfiquet) [inline client/src/Auth/Register/register.test.js]: I'm confused by this. This is the Register page, I don't think there is any need to initialise the input fields to any previous values. Or is that to test that you can reset the fields? Because it's not testing that.

What this file should be doing is test that the user interaction is working as expected, e.g.:
- it('should show empty fields for the user name and password with submit button disabled and no confirm password field')
- it('should show the username error message when the username input string is shorter than 4 characters')
- it('should not show the username error message when the username input string is 4 or more characters')
etc, etc...

It's ok to only write the descriptions of the tests without the implementation in the first instance, like I did above. This page probably needs a long list of tests and I'm not too bothered if you can't write them all now. But I would really like all the expected behaviour to be described. It's documentation and it allows someone else to write the tests if they have time.

Comment (Heyjp) [reply client/src/Auth/Register/register.test.js]: It's also a controlled component, it's there to make sure the inputs reflect the state of the parent container. Will get on and do the rest of the tests.

File: client/src/Auth/Register/register.js [hunk 8]
```diff
@@ -60,7 +63,7 @@ const Register = (props) => {
                             onBlur={props.onBlur}
                             onChange={(e) => {props.handleChange('password', e)} }
                         />
-                        {focus === 'password' && errors.password && <p className="error-text">Passwords must be over 6 characters</p>}
+                        {focus === 'password' && errors.password && <p className="error-text">Passwords must be over 5 characters</p>}
```
Comment (sfiquet) [inline client/src/Auth/Register/register.js]: This should also use constant PASSWORD_LENGTH_MIN. This way when we change the minimum length, the message will stay correct.

**Comment by Heyjp** [comment]
Any tips on how I can specifically test for that portrait tablet styling? Not sure if you have noticed, not the most adept at responsive design. 😅

**Comment by sfiquet** [comment]
You don't specifically test for portrait tablet, you test for width. 

In general, you start styling for smallest width (i.e. portrait mobile and desktop window resized to super-small width). That's also the default for Bootstrap if you don't specify anything (xs is default). When you're happy with it, you resize your window  horizontally and you stop when it starts looking weird. If you do it in Chrome devtools, you can see the window dimensions as you do it. For that width you add a `min-width` media query to adjust the style. Then you do it again until it looks good at all widths.

With Bootstrap you can use their predefined classes instead of writing media queries (like you've been doing: sm for landscape phone onwards, md for tablets and bigger, etc...: I always refer to https://getbootstrap.com/docs/4.0/layout/overview/#responsive-breakpoints because I always forget what they stand for).

In this case maybe it looks suddenly narrow because you're jumping from `col-sm-8` to `col-md-6`? Not sure. I think I'll run the code again to see. It might also be possible to simplify the styling but I can't be sure until I try it. One thing I do when I style is edit the styling in devtools until it looks correct, then I change the styling in the code. It allows me to tinker and learn what works. Looking for a specific class in `bootstrap.css` also helps when you're not sure what it does exactly.

**Comment by sfiquet** [comment]
I think it works and looks great now. Just a tiny problem left: Some file names are not what they should be and it doesn't compile:
- `register.js` should be `Register.js` 
- `register.scss` should be `Register.scss`
- After renaming `register.scss` to `Register.scss`, for some reason, if you don't delete `register.css` manually, sass doesn't create `Register.css` and then React can't find it. Not sure if it's a bug in the library or if it's a config issue.
- Test file names don't break anything but they should be renamed to match their tested file:
  * `register.test.js` => `Register.test.js`
  * `auth.test.js` => `Auth.test.js`

**Comment by Heyjp** [comment]
Turns out if you capitalize a file name that isn't enough for git to track it. Had to rename twice in order to get it to register. Hopefully it is all good now.

File: client/src/Auth/Register/Register.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    const isDirty = checkDirty(focus, props.inputs, errors);
+    const isEnabled = Object.keys(errors).some(e => errors[e]);
+
+    return (
+        <div className="container-fluid d-flex flex-column my-auto">
+                <div className="d-flex flex-column col-xl-4 col-lg-5 col-md-6 col-sm-8 col-9 mx-auto">
+                    <div className="align-center pt-2 mb-3">
+                        <h2 className="register-text">Register</h2>
+                    </div>
+                    <img src={avatar} className="align-self-center my-3" width="120px" />
```
Comment (sfiquet) [inline client/src/Auth/Register/Register.js]: Please add an alt text to make the ESLint warning disappear. I think `alt=""` will suffice since the image is just decorative.
